test/test_modules.py

```diff
@@ -0,0 +1,41 @@
+from copy import deepcopy
+from torch.testing._internal.common_device_type import instantiate_device_type_tests
+from torch.testing._internal.common_modules import module_db, modules
+from torch.testing._internal.common_utils import TestCase, run_tests, freeze_rng_state
+
+
+class TestModule(TestCase):
+    _do_cuda_memory_leak_check = True
+    _do_cuda_non_default_stream = True
+
+    @modules(module_db)
+    def test_forward(self, device, dtype, module_info):
+        module_cls = module_info.module_cls
+        module_inputs = module_info.module_inputs_func(module_info, device=device, dtype=dtype,
+                                                      requires_grad=False)
+        for module_input in module_inputs:
+            if module_input.forward_input is None:
+                continue
+
+            with freeze_rng_state():
+                # === Instantiate the module. ===
+                args, kwargs = module_input.constructor_input.args, module_input.constructor_input.kwargs
+                m = module_cls(*args, **kwargs)
+                m.to(device).to(dtype)
+
+                # === Do forward pass. ===
+                args, kwargs = module_input.forward_input.args, module_input.forward_input.kwargs
+                outputs = m(*args, **kwargs)
+
+                # === Compare outputs to a reference if one is specified. ===
+                # TODO: Handle precision
+                reference_fn = module_input.reference_fn
```

Comment:
Interesting that the reference_fn is on the module input

Reply:
`ModuleInput` is possibly a bad name- it started out as a module analogue for `SampleInput`. We want the ability to have a reference function per (constructor args, forward args), so that was added there. I'm good with changing the name or whatever makes it less weird :p

Reply:
Names can be sorted later, this seems fine